crates/biome_fs/src/path.rs

```diff
@@ -3,34 +3,106 @@
 #[cfg_attr(
     feature = "serde",
     derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
 )]
-pub struct BiomePath {
-    path: PathBuf,
+// NOTE: The order of the variants is important, the one on the top has the highest priority
+pub enum FileKind {
+    /// A configuration file has the highest priority. It's usually `biome.json` and `biome.jsonc`
+    ///
+    /// Other third-party configuration files might be added in the future
+    Config,
+    /// It's usually `package.json` and `tsconfig.json`
+    Manifest,
+    /// An ignore file, like `.gitignore`
+    Ignore,
+    /// Files that are required to be inspected before handling other files.
+    ///
+    /// An example is the GraphQL schema
+    ToInspect,
+    /// A file to handle has the lowest priority. It's usually a traversed file, or a file opened by the LSP
+    #[default]
+    ToHandle,
 }
 
-impl Deref for BiomePath {
-    type Target = PathBuf;
+#[derive(Debug, Clone, Hash, Ord, PartialOrd, Eq, PartialEq, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct FileKinds(BitFlags<FileKind>);
 
-    fn deref(&self) -> &Self::Target {
-        &self.path
+impl From<FileKind> for FileKinds {
+    fn from(flag: FileKind) -> Self {
+        Self(BitFlags::from(flag))
     }
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Default)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize, schemars::JsonSchema)
+)]
+pub struct BiomePath {
+    /// The path to the file
+    path: PathBuf,
+    /// Determines the kind of the file inside Biome. Some files are considered as configuration files, others as manifest files, and others as files to handle
+    kind: FileKinds,
+    /// Whether this path (usually a file) was fixed as a result of a format/lint/check command with the `--write` filag.
+    was_written: bool,
+}
+
 impl BiomePath {
     pub fn new(path_to_file: impl Into<PathBuf>) -> Self {
+        let path = path_to_file.into();
+        let kind = path.file_name().map(Self::priority).unwrap_or_default();
+        Self {
+            path,
+            kind,
+            was_written: false,
+        }
+    }
+
+    pub fn new_written(path_to_file: impl Into<PathBuf>) -> Self {
+        let path = path_to_file.into();
+        let kind = path.file_name().map(Self::priority).unwrap_or_default();
         Self {
-            path: path_to_file.into(),
+            path,
+            kind,
+            was_written: true,
         }
     }
 
+    /// Creates a new [BiomePath], marked as fixed
+    pub fn to_written(&self) -> Self {
+        Self {
+            path: self.path.clone(),
+            kind: self.kind.clone(),
+            was_written: true,
+        }
+    }
+
+    pub fn was_written(&self) -> bool {
+        self.was_written
+    }
+
+    /// Adds a file kind to the current file
+    pub fn with_file_kind(mut self, kind: FileKind) -> Self {
+        self.kind.0.insert(kind);
```

Comment:
Do we want `Deref`/`DerefMut` implementations so we don't need to use the `.0` path in a bunch of places?